File: LeetCodeCPlusPlus/py47.cpp
// 全排列，有重复数字，不允许重复
#include <vector>
using namespace std;
class Solution {
    vector<int>tmp;
    vector<bool>used;
private:
    void dfs(vector<vector<int>>&result, vector<int>&nums, int index) {
         if (index==nums.size()) {
             result.push_back(tmp);
             return;
         }
        
        for (int i=0;i<nums.size();i++) 
            if (!used[i]) {
                if (i>0 && nums[i]==nums[i-1] && !used[i-1]) // 此处used[i-1]前加不加!代表两种情况，一种剪右支，一种剪左支
                    continue;
                used[i] = true;
                tmp.push_back(nums[i]);
                dfs(result, nums, index+1);
                tmp.pop_back();
                used[i] = false;
            }
        
    }
public:
    vector<vector<int>> permuteUnique(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        used.resize(nums.size());
        vector<vector<int>> result;
        dfs(result, nums, 0);
        return result;
    }
};

File: LeetCodePython/py47.py
# 全排列，有重复数字，不允许重复

class Solution(object):
    def permuteUnique(self, nums):
        result = []
        set_ = set()
        def backtrack(nums, tmp):
            if not nums:
                ans = ",".join(str(item)for item in tmp)
                if ans not in set_:
                    result.append(tmp)
                    set_.add(ans)
                return 
            for i in range(len(nums)):
                backtrack(nums[:i] + nums[i+1:], tmp + [nums[i]])
        backtrack(nums, [])
        return result